### PR TITLE
Rewrite specs

### PR DESCRIPTION
**Pull request by KevinBrowne**
The current specs are not exhaustive, failing to find a number of bugs in the gem.

Rewrite the specs to:

* test individual units rather than the gem as a whole at a high level

* clean up after themselves as much as possible

* have all data and setup for the specs in the spec files themselves

* test more scenarios

* draw attention to existing bugs that we're not fixing yet

* provide a more reliable basis to ensure the correct behaviour of future refactors

This commit also removes an unused private method from Sinject::Container, discovered when it showed as uncovered in the spec coverage report.

In the absence of CI, the specs were run locally and confirmed to pass under the following versions of MRI:

* 2.5.9 (previously used in CI)
* 2.6.7 (previously used in CI)
* 2.6.10
* 2.7.3 (previously used in CI)
* 2.7.8
* 3.0.6
* 3.1.4
* 3.2.3
* 3.3.0

### CODE REVIEW

File: spec/sinject/dependency_group_spec.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    # Call the GC after each example to try to clean up the
+    # definitions of the test classes.
+    #
+    # In some circumstances, something may hold a reference to a test
+    # class long enough to save it from the garbage collector in the
+    # after block. Force the GC both before and after to be sure.
+    before { ObjectSpace.garbage_collect }
+    after { ObjectSpace.garbage_collect }
+    # A final GC call, as if to say that we really mean it!
+    after(:all) { ObjectSpace.garbage_collect }
```
[the remark this answers is not in the document]
Comment (KevinBrowne) [reply spec/sinject/dependency_group_spec.rb:39]: It depends on the Ruby version and, I suspect, on timing.

An object won't be cleaned up on GC if there's a live reference to it. Unfortunately, it's hard to predict how long a given Ruby version may hold on to a reference to a class. These difficulties multiply when you throw more versions into the mix.

For example, under some versions, these tests fail without the GC in the `before` block. In some versions, accessing `instance.class` seems to cause a longer-lived reference. In some versions, it appears that if you name the class, you may never get rid of it. I even had GC failures once when I compared a class to itself! That's why the assertions themselves are very careful about what properties they assert on.

Basically, the `before` and `after` blocks are both necessary in some Ruby versions. The `after(:all)` may or may not be needed. It _is_ a bit of prayer: it might not be needed and might not do anything but sometimes it might be answered when needed, and it shows my devotion to cleaning up after myself. It's a small, harmless devotion to propitiate the interpreter.

I take this all as a sign that trawling `ObjectSpace` to find the dependency groups is not the best approach.

File: spec/sinject/container_spec.rb [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+            def with_args(first, second, third = true, fourth, fifth:, sixth: true); end
+          end
+
+          expect { container.register(key: :foo, class: klass, contract: contract) }
+            .to raise_error Sinject::DependencyContractInvalidParametersException
+        end
+      end
+
+      context 'and an optional positional parameter in the contract is made required in the class' do
+        it 'rasises DependencyContractInvalidParametersException' do
```
[the remark this answers is not in the document]
Comment (KevinBrowne) [reply spec/sinject/container_spec.rb:217]: Well, that's annoying! Why do I even bother running a spell-checker if I don't pay attention to it? 🤦🏻‍♂️

**Review by timlapluie — APPROVED** [review]
